### PR TITLE
New Pair

### PR DESCRIPTION
**Pull request by Kalzak**
Agent implementation for monitoring emission of `pairCreated` events from `Swap Factory V1`

### CODE REVIEW

File: impossible-finance/New-Pair/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+## Alerts
+
+Describe each of the type of alerts fired by this agent
+
+- [agentID tbd]
+  - Fired when the contract `Swap Factory V1` emits a `PairCreated` event
+  - Severity is always set to "info"
+  - Type is always set to "info"
+  - Mention any other type of metadata fields included with this alert
```
Comment (CeliktepeMurat) [inline impossible-finance/New-Pair/README.md:19]: this line should be removed

File: impossible-finance/New-Pair/package.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "forta-agent-starter",
```
Comment (CeliktepeMurat) [inline impossible-finance/New-Pair/package.json:2]: please add agent name here

File: impossible-finance/New-Pair/package.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  "version": "0.0.1",
+  "description": "Forta Agent Typescript starter project",
```
Comment (CeliktepeMurat) [inline impossible-finance/New-Pair/package.json:4]: also description of agent here

**Review by CeliktepeMurat — CHANGES_REQUESTED** [review]
left minor requests

File: impossible-finance/New-Pair/src/agent.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    return Finding.fromObject({
+      name: 'New pair created',
+      description: 'A new pair has been created in Swap Factory V1',
+      alertId: alertId,
+      severity: FindingSeverity.Info,
+      type: FindingType.Info,
+      protocol: 'Impossible Finance',
+      metadata: {
+        token_0: metadata!.topics[1],
+        token_1: metadata!.topics[2],
```
Comment (e1Ru1o) [inline impossible-finance/New-Pair/src/agent.ts:37]: This topics must be decoded.

File: impossible-finance/New-Pair/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    let token0Address = '0xab5801a7d398351b8be11c439e05c5b3259aec9c';
+    let token1Address = '0xab5801a7d398351b8be11c439e05c5b3259aec9d';
+
+    const tx: TransactionEvent = new TestTransactionEvent()
+      .addEventLog(
+        SWAP_FACTORY_IFACE.getEvent('PairCreated').format('sighash'),
+        TARGET,
+        '',
+        token0Address,
+        token1Address,
```
Comment (e1Ru1o) [inline impossible-finance/New-Pair/src/agent.spec.ts:78]: Topics must be encoded

File: impossible-finance/New-Pair/src/agent.spec.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        SWAP_FACTORY_IFACE.getEventTopic('PairCreated'),
```
Comment (e1Ru1o) [inline impossible-finance/New-Pair/src/agent.spec.ts:79]: this line must be removed

**Review by e1Ru1o — CHANGES_REQUESTED** [review]
Two suggestions:
- Fill all the events data in each of the tests (i.e `topics` & `data`) just for consistency
- Avoid using `provideEventChecker` form `forta-agent-tools` because this template uses `txEvent.filterEvent` and that function is being deprecated. Instead of that try `txEvent.filterLog`

Also I would like to see a tests that detect multiple findings from the same transaction

File: impossible-finance/New-Pair/src/agent.spec.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  it('should ignore different event from same contract', async () => {
+    const tx: TransactionEvent = new TestTransactionEvent()
+      .addEventLog(
+        'IrrelevantEvent(address)',
+        factory,
+        '',
+        token0,
+        token1,
+        SWAP_FACTORY_IFACE.getEventTopic('PairCreated'),
+      );
```
Comment (e1Ru1o) [inline impossible-finance/New-Pair/src/agent.spec.ts:76]: Please at fill the event with data & topics accordingly with the signature you are providing

**Review by e1Ru1o — CHANGES_REQUESTED** [review]

File: impossible-finance/New-Pair/src/agent.spec.ts (hunk 7 above)
Comment (Kalzak) [reply impossible-finance/New-Pair/src/agent.spec.ts:76]: This has been addressed in the latest commit

File: impossible-finance/New-Pair/package.json [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    "file": "npm run build && forta-agent run --file",
+    "publish": "forta-agent publish",
+    "push": "forta-agent push",
+    "disable": "forta-agent disable",
+    "enable": "forta-agent enable",
+    "keyfile": "forta-agent keyfile",
+    "test": "jest"
+  },
+  "dependencies": {
+    "bignumber.js": "^9.0.1",
```
[the remark this answers is not in the document]
Comment (Kalzak) [reply impossible-finance/New-Pair/package.json]: This has been addressed in the latest commit

**Review by e1Ru1o — APPROVED** [review]

**Pull request merged**